internal/clientconn/conninfo/conn_info.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package conninfo provides a ConnInfo struct that is used to handle connection-specificinfo
+// and can be shared through context.
+package conninfo
+
+import (
+	"context"
+	"net"
+)
+
+// contextKey is a special type to represent context.WithValue keys a bit more safely.
+type contextKey string
+
+// connInfoKey stores the key for withConnInfo context value.
+const connInfoKey = contextKey("connInfo")
+
+// ConnInfo represents connection info.
+type ConnInfo struct {
+	PeerAddr net.Addr
+}
+
+// WithConnInfo returns a new context with the given ConnInfo.
+func WithConnInfo(ctx context.Context, connInfo *ConnInfo) context.Context {
+	return context.WithValue(ctx, connInfoKey, connInfo)
+}
+
+// GetConnInfo returns the ConnInfo value stored in ctx, or empty connInfo if there is nothing stored there.
+func GetConnInfo(ctx context.Context) *ConnInfo {
+	value := ctx.Value(connInfoKey)
+	if value == nil {
+		return &ConnInfo{}
```

Comment:
I would say let's panic in that case as it should not happen

Reply:
I disagree but commit :)

Reply:
My point is that the value will be absent only if we, developers, forget to add `WithConnInfo` to all places where it should be. It is not possible to trigger this error by any user action/input. Therefore, that's a programmer's error. Therefore, that should be a panic. (And that panic will not crash the whole program as we recover from them at the connection level, but it will be very noticeable for users who would create an issue or for us.)

Reply:
Yes, I understand that point. But then should we go deeper and check that the structure is not a zero value structure, the address is actually set and not nil etc?.. Probably, not (preventing missing context seems to be enough), but it's just a bit hard to differentiate one dev mistake from another.